### PR TITLE
Add support for concurrent host IO and compute

### PR DESCRIPTION
**Pull request by tdb-alcorn**
This change makes it possible to parallelize data moves and matrix multiplies by moving all host IO to the secondary memory port of the local cache. We introduce a host router to handle multiplexing the dual DRAM ports into that memory port, and remove host IO ports from the existing router. This has the incidental benefit of adding support for the memory=>DRAM1 data move operation.

### CODE REVIEW

File: rtl/src/main/scala/tensil/formal/tcu/RouterFormal.scala [hunk 1]
```diff
@@ -26,52 +26,40 @@ class RouterFormal extends Formal {
...
-  val memInputFromHost = Node(
-    m.io.mem.input,
-    filter = m.io.control.bits.kind === DataFlowControl.dram0ToMemory
-  )
   val memOutputForAcc = Node(
     m.io.mem.output,
     filter =
-      m.io.control.bits.kind === DataFlowControl.memoryToAccumulator ||
-        m.io.control.bits.kind === DataFlowControl.memoryToAccumulatorAccumulate
+      m.io.control.bits.kind === DataFlowControl.memoryToAccumulator //||
```
Comment (petrohi) [inline rtl/src/main/scala/tensil/formal/tcu/RouterFormal.scala:41]: `//||` can be removed.

File: rtl/src/main/scala/tensil/formal/tcu/TCUFormal.scala [hunk 2]
```diff
@@ -43,7 +43,7 @@ class TCUFormal extends Formal {
 
   // v.cover(io.instruction.bits.opcode === Opcode.DataMove)
   // when(io.instruction.bits.opcode === Opcode.DataMove) {
-  //   v.assume(DataFlowControl.isValid(io.instruction.bits.flags))
+  //   v.assume(DataMoveKind.isValid(io.instruction.bits.flags))
   //   val args = Wire(
   //     new DataMoveArgs(layout)
   //   )
```
Comment (petrohi) [inline rtl/src/main/scala/tensil/formal/tcu/TCUFormal.scala:49]: Do we need this block of code?

File: rtl/src/main/scala/tensil/tcu/instruction/DataMove.scala [hunk 3]
```diff
@@ -52,12 +52,46 @@ object DataMoveArgs {
...
 }
+
+object DataMoveKind {
+  type DataMoveKind = UInt
+
+  val dram0ToMemory = 0x0.U
+  val memoryToDram0 = 0x1.U
+  val dram1ToMemory = 0x2.U
+  val memoryToDram1 = 0x3.U
+  // 0x4 - 0xb are unused
```
Comment (petrohi) [inline rtl/src/main/scala/tensil/tcu/instruction/DataMove.scala:73]: Does decoupling instruction flags and flow control mean we don't need this unused gap anymore?

File: rtl/src/main/scala/tensil/tcu/DataFlowControl.scala [hunk 4]
```diff
@@ -31,27 +31,12 @@ object DataFlowControlWithSize {
 }
 
 object DataFlowControl {
```
Comment (petrohi) [inline rtl/src/main/scala/tensil/tcu/DataFlowControl.scala:33]: Maybe `LocalDataFlowControl` and `LocalRouter` to clarify the difference with `HostDataFlowControl` and `HostRouter`?

**Review by petrohi — APPROVED** [review]
Tested with ResNet20 simulation and synthesized for Ultra96. Checks out with no changes in performance. Few small nitpicks in the comments. LGTM!

File: rtl/src/main/scala/tensil/formal/tcu/RouterFormal.scala (hunk 1 above)
Comment (tdb-alcorn) [reply rtl/src/main/scala/tensil/formal/tcu/RouterFormal.scala:41]: Done.

File: rtl/src/main/scala/tensil/formal/tcu/TCUFormal.scala (hunk 2 above)
Comment (tdb-alcorn) [reply rtl/src/main/scala/tensil/formal/tcu/TCUFormal.scala:49]: We do not!

File: rtl/src/main/scala/tensil/tcu/instruction/DataMove.scala (hunk 3 above)
Comment (tdb-alcorn) [reply rtl/src/main/scala/tensil/tcu/instruction/DataMove.scala:73]: Correct, but I think it's fine to have the external and internal datamove kinds live at opposite ends of the allowable range. Also, this maintains backwards compatibility of existing binaries. I'm open to alternatives though!

File: rtl/src/main/scala/tensil/tcu/DataFlowControl.scala (hunk 4 above)
Comment (tdb-alcorn) [reply rtl/src/main/scala/tensil/tcu/DataFlowControl.scala:33]: Done.

File: rtl/src/main/scala/tensil/tcu/instruction/DataMove.scala (hunk 3 above)
Comment (petrohi) [reply rtl/src/main/scala/tensil/tcu/instruction/DataMove.scala:73]: Good point. Let's keep it then.

**Pull request merged**